src/Compiler/Mapper/Object/MapDiscriminatedObject.php

```diff
@@ -0,0 +1,160 @@
+                    ),
+                ),
+            ]),
+        ];
+
+        $isDiscriminatorPresent = $builder->funcCall($builder->importFunction('array_key_exists'), [$builder->val($this->discriminatorFieldName), $value]);
+        $isDiscriminatorMissing = $builder->not($isDiscriminatorPresent);
+
+        $statements[] = $builder->if($isDiscriminatorMissing, [
+            $builder->throw(
+                $builder->staticCall(
+                    $builder->importClass(MappingFailedException::class),
+                    'missingKey',
+                    [$path, $this->discriminatorFieldName],
+                ),
+            ),
+        ]);
+
+        $discriminatorRawValue = $builder->arrayDimFetch($value, $builder->val($this->discriminatorFieldName));
+        $discriminatorPath = $builder->arrayImmutableAppend($path, $builder->val($this->discriminatorFieldName));
+        $discriminatorMapperMethodName = $builder->uniqMethodName('map' . ucfirst($this->discriminatorFieldName));
+        $discriminatorMapperMethod = $builder->mapperMethod($discriminatorMapperMethodName, new MapString())->makePrivate()->getNode();
+        $discriminatorMapperCall = $builder->methodCall($builder->var('this'), $discriminatorMapperMethodName, [$discriminatorRawValue, $discriminatorPath]);
+        $builder->addMethod($discriminatorMapperMethod);
+
+        $validMappingKeys = array_keys($this->subtypeMapping);
+        $isDiscriminatorValid = $builder->funcCall($builder->importFunction('in_array'), [$discriminatorRawValue, $builder->val($validMappingKeys), $builder->val(true)]);
+
+        $expectedDescription = $builder->concat(
+            'one of ',
+            $builder->funcCall($builder->importFunction('implode'), [
+                ', ',
+                $builder->val($validMappingKeys),
+            ]),
+        );
+
+        $statements[] = $builder->if($builder->not($isDiscriminatorValid), [
+            $builder->throw(
+                $builder->staticCall(
+                    $builder->importClass(MappingFailedException::class),
+                    'incorrectValue',
+                    [$discriminatorRawValue, $discriminatorPath, $expectedDescription],
+                ),
+            ),
+        ]);
+
+        $subtypeMatchArms = [];
+
+        foreach ($this->subtypeMapping as $key => $subtype) {
+            $mapperProviderMethodCall = $builder->methodCall($provider, 'get', [
+                $builder->classConstFetch($builder->importClass($subtype), 'class'),
+            ]);
+
+            $subtypeMatchArms[] = $builder->matchArm(
+                $builder->val($key),
+                $builder->methodCall($mapperProviderMethodCall, 'map', [$value, $path]),
+            );
+        }
+
+        $subtypeMatchArms[] = $builder->matchArm(
+            null,
+            $builder->throwExpr(
+                $builder->new(
+                    $builder->importClass(LogicException::class),
```

Comment:
You can throw the `MappingFailedException::incorrectValue()` here.

Reply:
Replaced